### PR TITLE
feat(common/models): support direct-child access for Trie node iteration 📚 

### PR DESCRIPTION
**Pull request by jahorton**
This PR was originally part of https://github.com/keymanapp/keyman/pull/10973.

These changes will facilitate the next PR's changes (#11870), in which it becomes possible to support prediction, in full, solely via the `LexiconTraversal` iterator structure.

Note that this gives us a neat way to support new model types:  if a `LexiconTraversal` can be defined for that model, we can efficiently search it and predict for it, even if it's not actually supported by a Trie!  There would be no need for the model to define a separate `predict` method on the model itself, as it can be implemented via the traversal.

@keymanapp-test-bot skip

### CODE REVIEW

File: common/models/templates/src/trie-model.ts [hunk 1]
```diff
@@ -100,13 +100,62 @@ class Traversal implements LexiconTraversal {
    */
   totalWeight: number;
 
-  constructor(root: Node, prefix: string, maxWeight: number) {
+  constructor(root: Node, prefix: string, totalWeight: number) {
     this.root = root;
     this.prefix = prefix;
-    this.totalWeight = maxWeight;
+    this.totalWeight = totalWeight;
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:106]: I wonder if we can `Object.assign(this, {root, prefix, totalWeight});`? Syntactic prettiness only... dunno if there are downsides?

File: common/models/templates/src/trie-model.ts [hunk 2]
```diff
@@ -100,13 +100,62 @@ class Traversal implements LexiconTraversal {
...
+
+  child(char: USVString): LexiconTraversal | undefined {
+    /*
+      Note: would otherwise return the current instance if `char == ''`.  If
+      such a call is happening, it's probably indicative of an implementation
+      issue elsewhere - let's signal now in order to catch such stuff early.
+    */
+    if(char == '') {
+      return undefined;
+    }
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:117]: Do we want to log something like this?

File: common/models/templates/src/trie-model.ts [hunk 3]
```diff
@@ -100,13 +100,62 @@ class Traversal implements LexiconTraversal {
...
+
+    // Split into individual code units.
+    let steps = char.split('');
+    let traversal: ReturnType<Traversal["_child"]> = this;
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:121]: This is a bit funky -- I had to read it a few times to try and figure out what that was doing...

File: common/models/templates/test/test-trie-traversal.js [hunk 4]
```diff
@@ -101,6 +105,38 @@ describe('Trie traversal abstractions', function() {
     assert.isEmpty(eKeys);
   });
 
+  it('direct traversal with simple internal nodes', function() {
```
Comment (mcdurdin) [inline common/models/templates/test/test-trie-traversal.js:108]: The description is not very fluent -- ideally, should be something that works with "it", like:

```suggestion
  it('directly traverses a trie model with simple internal nodes', function() {
```

But this is a nit only, I don't really care that much

File: common/models/types/index.d.ts [hunk 5]
```diff
@@ -50,6 +50,23 @@ declare interface LexiconTraversal {
    */
   children(): Generator<{char: USVString, traversal: () => LexiconTraversal}>;
 
+  /**
+   * Allows direct access to the traversal state that results when appending a
+   * `char` representing a single UTF-16 codepoint to the current traversal
```
Comment (mcdurdin) [inline common/models/types/index.d.ts:55]: `single UTF-16 codepoint` is a little confusing on first read to me. Do you mean a `single codepoint encoded in UTF-16` (i.e. up to 2 UTF-16 code units?) 

I am not sure that `USVString` is really a thing in vanilla JavaScript/Typescript?

**Review by mcdurdin — APPROVED** [review]
LGTM. Nits only. Although use of `USVString` could be worthy of deeper discussion.

File: common/models/templates/src/trie-model.ts (hunk 1 above)
Comment (jahorton) [reply common/models/templates/src/trie-model.ts:106]: TypeScript doesn't seem to perform type-checking well when using the suggested pattern.  I feel that this downside is too significant, so I'd rather not change to using it here.

File: common/models/types/index.d.ts (hunk 5 above)
Comment (jahorton) [reply common/models/types/index.d.ts:55]: I actually mean "one or more codepoints", which can be seen in #11870.  I'll rework to something that incorporates both this and that.

> I am not sure that USVString is really a thing in vanilla JavaScript/Typescript?

I'm not either, but we tagged a lot of other things that use strings within `common/models/types/index.d.ts` with USVString.  Just being consistent with "what is".  I never really understood the reason for the distinction being made there myself.

File: common/models/templates/src/trie-model.ts (hunk 2 above)
Comment (jahorton) [reply common/models/templates/src/trie-model.ts:117]: It's hard for the mobile apps to properly log console-logging statements made from the WebWorker.  I don't know that it's worth the trouble when we don't already have a pattern established for capturing said logs.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.70-alpha](https://keyman.com/downloads/releases/alpha/18.0.70)

File: common/models/types/index.d.ts (hunk 5 above)
Comment (mcdurdin) [reply common/models/types/index.d.ts]: * https://github.com/keymanapp/keyman/issues/12046

**Pull request merged**